### ISSUE TITLE
We made a toolkit can parallelize almost all the Hugging Face models. But we have some question !

### ISSUE THREAD

**Issue by hyunwoongko** [huggingface/transformers#12772]
We recently developed an opensource called `parallelformers,` (https://github.com/tunib-ai/parallelformers) and have a few questions, so we write an issue here.

Q. As a logo, an image homage to the hugging face was used. Not exactly the same CI, but from Unicode. Will it be a problem?

Q. What do you think about collaboration? We can include model parallelization for all models in hugging face transformers.

---

The following is what I posted on Reddit to promote our opensource.

Hello, I am writing to inform you about the release of Parallelformers (https://github.com/tunib-ai/parallelformers), a model parallelization library at TUNiB. Parallelformers is a toolkit that supports inference parallelism for 68 models in Huggingface Transformers with 1 line of code.

Previously, DeepSpeed-Inference was used as a parallelization toolkit for model inference.

(1) It was impossible to deploy to the web server due to the process flow,

(2) Lack of integration with Huggingface Transformers, which has now become the de facto standard for natural language processing tools. (DeepSpeed-Inference only supports 3 models)

(3) Also, since parallelization starts in the GPU state, there was a problem that all parameters of the model had to be put on the GPU before parallelization.

Parallelformers solved a number of problems in DeepSpeed-Inference. Using this toolkit internally, we were able to easily deploy a large model to our web server, reducing the cost of deployment by up to 3-5x. More detailed information and source code can be found on GitHub. Thanks !

**Comment by LysandreJik** [comment]
Hello @hyunwoongko, thanks a lot for sharing, this is a really cool project! No problem at all regarding the image homage (really cool logo by the way!)

I'm pinging @stas00 who has led the efforts of model parallelization and DeepSpeed integration on our side and would probably be interested. Also pinging @sgugger as he has done some similar work.

**Comment by stas00** [comment]
Thank you for implementing and sharing your project, @hyunwoongko,

I haven't had a chance to study your project closely yet, but is it correct that you implemented tensor parallelism from Megatron?
In other words this outstanding feature: https://github.com/huggingface/transformers/issues/10321 except for inference-only?

There are many types of model parallelism and it is much easier to understand things when the generic MP term is not used, but an explicit type is described. Here is my initial attempt to create the distinctions https://huggingface.co/transformers/master/parallelism.html

**Comment by hyunwoongko** [comment]
I am not good at English, so I wrote this article with the help of a google translator. If you find any strange expressions, please let me know. In order to provide you with more accurate information, I will write a long explaination.

---

I think our implementation is an extension of Hugging Face Transformers. So, I'd like to take this implementation a bit further and discuss making it a default feature of Hugging Face Transformers. I will share with you the problem that I solved to make parallelization of almost models of hugging face possible.

**Comment by hyunwoongko** [comment]
## 1. Solving problem about parallelization scalability using "only" Huggingface Transformers
Firstly, I inspired by DeepSpeed-Inference's parallelization codes and implemented using Megatron tensor slicing method like them. you can check our megatron implementations here: https://github.com/tunib-ai/parallelformers/blob/main/parallelformers/parallel/replacing.py#L291 

Currently, DeepSpeed-Inference's parallelization is implemented as follows:

1) loading the weights from huggingface model
2) slice weight and bias tensors (by megatron way)
3) replace all the original transformer layers to megatron layers

And my implementation is the same. But the main difference is that they replace huggingface layers to their own CUDA fused kernel. This makes model fast, but it has many limitations. (you can check their implementation here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/ops/transformer/inference/transformer_inference.py)

![image](https://user-images.githubusercontent.com/38183241/126205396-3d2c9e8a-2275-486d-b9eb-e31cfd3c1300.png)

Their biggest limitation is the parallelization scalability of the models. And I remember that Huggingface also experienced this problem. The existing layer-wise parallelization method of GPT2 and T5 is very poorly scalable because we have to fix the code for every model.

The implementation of DeepSpeed-Inference is very poorly scalable because it must use the CUDA kernel. If they need BigBird's random attention they have to implement it in CUDA. Also, if they need GPTNeo's local attention, they have to implement it with CUDA too. They need to implement all the model specific mechanism using CUDA. 

It is huge limitation. And currently, They don't support the seq2seq architecture and only support 3 Huggingface Transformers architectures. (Bert, GPTNeo, GPT2) (you can check here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/module_inject/replace_policy.py)

![image](https://user-images.githubusercontent.com/38183241/126205417-efe451f6-ca8a-4c4e-b946-d92eb8e64727.png)

![image](https://user-images.githubusercontent.com/38183241/126205657-f366d2bf-b051-435e-9bb6-0956359fc04d.png)

But My implementation performs only parallelization while keeping the code of the hugging face's `forward` method. Although slower than the fused kernel, we were able to successfully parallelize almost all models in the hugging face. (Vision models such as ViT and CLIP and speech models such as Wav2Vec2 can also be parallelized) 

As a result of the experiment, it was possible to parallelize 68 models out of a total of 70 models. you can check supported model list here: https://tunib-ai.github.io/parallelformers/intro/SUPPORTED_MODELS.html and you can test using these codes: https://github.com/tunib-ai/parallelformers/tree/main/tests

**Comment by hyunwoongko** [comment]
## 2. Solving problem about parallel deployment

Second, DeepSpeed is currently not available for deployment. Their launcher creates multiple subprocesses using user code file. These implementations are usually used for training (This method is generally used when training with DDP.) 

![image](https://user-images.githubusercontent.com/38183241/126204577-85b35ec6-9f40-4d7f-ad8c-08cddfa44177.png)

The problems with these implementations are: 

![image](https://user-images.githubusercontent.com/38183241/126206086-3b91c22e-5220-43e6-8cc4-9aaf6a665c9e.png)

1. The user's code is executed multiple times at the same time. Therefore, the web server port overlaps and the hugging face model is downloaded multiple times from `from_pretrained`, exceeding the CPU memory. I recently uploaded a bunch of big models to the Hugging Face Hub to check this out. (https://huggingface.co/hyunwoongko/blenderbot-9B, https://huggingface.co/hyunwoongko/megatron-11B) Even though I used more than 200GB of RAM, the memory exceeds when I ran the Deep Speed Launcher and downloaded BlenderBot 9B with a total of 8 GPUs at the same time.

2. Joining of processes is not possible. A parent process can join a child process. However, control of the parent process is not given to the user. Therefore, deperalleization cannot be performed in this way.

To solve this, we devised the idea called 'inversion of process control'.

![image](https://user-images.githubusercontent.com/38183241/126204917-25e42e27-9f3b-4219-92c8-205d5a96ea1b.png)

We have implemented some mutex handling and succeeded in transferring data using the producer-consumer pattern.

![image](https://user-images.githubusercontent.com/38183241/126205342-b983daf4-aba6-453a-a57e-d8e4a8068382.png)

We developed this deployment specific launcher, which solved all problems with inference and deployment. Of course, using this launcher is optional. If necessary, we can apply our parallelization method to training using the same method as deep speed (like DDP).

**Comment by hyunwoongko** [comment]
## 3. Solving Memory Inefficiency

There are several memory inefficient implementations of DeepSpeed. They upload all parts of the model to the GPU before starting parallelization. If this were possible, there would be no need to parallelize the model.

![image](https://user-images.githubusercontent.com/38183241/126210339-03bb1f2e-3695-415a-9c6d-f6f3980cc543.png)

To solve this problem, we changed the parallelization order as follows.

![image](https://user-images.githubusercontent.com/38183241/126210377-7246a489-dae3-42d2-8c0f-dff845ec0375.png)

This change allowed us to efficiently parallelize very large models.

**Comment by hyunwoongko** [comment]
These are the ideas we came up with and implemented. I want to discuss about collaboration . If you apply our method, we can parallelize all models of hugging face very easily.

We provide a kind of configuration object called `Policy` for each model. https://github.com/tunib-ai/parallelformers/tree/main/parallelformers/policies Parallelization is possible just by adding a file called `policy_bert.py` like `modeling_bert` and `tokenization_bert` to the existing hugging face. Most of the code of the existing model does not need to be modified.

Here's an example of policy: https://github.com/tunib-ai/parallelformers/blob/main/parallelformers/policies/bert.py
And check out more information here: https://tunib-ai.github.io/parallelformers/intro/POLICY.html

**Comment by hyunwoongko** [comment]
@stas00 These are the ideas we came up with and implemented. I want to discuss about collaboration with these mechanisms. If you apply our method, we can parallelize all models of hugging face very easily.

We provide a kind of configuration object called Policy for each model. https://github.com/tunib-ai/parallelformers/tree/main/parallelformers/policies Parallelization is possible just by adding a file called policy_bert.py like modeling_bert.py and tokenization_bert.py to the existing hugging face. Most of the code of the existing model does not need to be modified.

Here's an example of policy: https://github.com/tunib-ai/parallelformers/blob/main/parallelformers/policies/bert.py
And check out more information here: https://tunib-ai.github.io/parallelformers/intro/POLICY.html

Currently our framework is designed to support inference only. However, with a few modifications to the code, I think it will be possible to train as well. To do that, we will have to write 2D parallelism code considering the combination with data parallelism. If you want to provide a training feature, we can implement it together. I think it is better to develop it further and mount it on the hugging face transformers rather than as it is now.

I don't have any personal desires, I just want more people to use the large-scale model more easily. (My motto has always been "NLP for everyone.") Currently, it is too difficult to use a large-scale model (e.g. T5, Blenderbot). I want to solve this problem, and I think that collaborating with you will make it easier for more people to use the large-scale model.

**Comment by hyunwoongko** [comment]
These are additional figures to explain more accurate

## Inversion of Process control
![image](https://user-images.githubusercontent.com/38183241/126270409-e60ca501-a3a8-4426-9566-47d7a231d0ba.png)

## Data communication via shared memory
![image](https://user-images.githubusercontent.com/38183241/126270420-3c6c826c-4376-4543-93c2-bf08af0224a2.png)

![image](https://user-images.githubusercontent.com/38183241/126270480-a1b5cd0a-9c26-4061-a831-c3a3546e0ef4.png)

## Method hijacking (for easy usage)
![image](https://user-images.githubusercontent.com/38183241/126270612-cbbbcd8f-0005-4b5a-bb9a-7f4d56365841.png)

**Comment by hyunwoongko** [comment]
@stas00 
I'll write a blog post about the architecture of our tools soon and share them.

**Comment by stas00** [comment]
Oh, but why have you deleted all the detailed comments you posted earlier? I was looking forward to studying those and now they are all gone.

My plan was to do a feasibility study and then see if we can integrate your work into HF transformers.

**Comment by hyunwoongko** [comment]
Because the article was written too hastily and too long, I decided that it would be more helpful for you to understand it by organizing it more neatly and accurately than explaining it in the issue comments. (I was going to blog soon, maybe within this week.)

**Comment by stas00** [comment]
at the end I was able to cheat since github sent me all the comments ;) So I have just read those.

It wasn't long at all, on the contrary I'd say it could use more details in places.

And I'm very appreciating you wanted to merge this into HF transformers! That would be an amazing contribution!

So bottom line, beside the leaner launcher, the core of your project is Tensor Parallel from built upon Megatron-LM, correct? this is exactly what I was planning to work on when I had free time, so your timing is perfect.

Let's discuss the training side of it. I think to most users of HF transformers that would be the most important application of Tensor parallelism. So in the deleted note you mentioned that DDP-support needs to be integrated to make it work in training. That's the MPU part, right? And we probably should think about Pipeline too while building the MPU, while not implementing it just yet.

Also do you think it'd be a good idea to invite @RezaYazdaniAminabadi into this process, so that gradually we can use your project's flexibility and add Deepspeed CUDA kernels speeds where possible. i.e. work together with the Deepspeed project. That's of course if Reza is interested and his superiors support the effort. We already discussed with Deepspeed to start deploying some of their kernels in the transformers (but haven't done anything yet). 

How do you propose we work on integrating this? Perhaps pick a few models first and work on a PR that integrates those and then in a subsequent PR work on other models? Probably leaving the optional launcher out at first and then considering it next?

On a personal note: we are about to launch the first training of the Big Science project https://github.com/bigscience-workshop/ so my availability depends on that, if when we launch it all goes well, I will have more time, if not please bear with me, but I will do my best to support this integration process at least a bit at a time.

If you have any questions or concerns please don't hesitate to ask. I will try to address those.

**Comment by hyunwoongko** [comment]
I have sent my thoughts about collaboration to your email (stas@stason.org) !

**Comment by stas00** [comment]
Thank you for emailing me your notes, @hyunwoongko 

We need to discuss it here and not in private, since this is not my personal project. Therefore please re-paste all or just the parts that you feel are open to the public and we will continue the discussion here.

**Comment by hyunwoongko** [comment]
Okay. First of all, I'm very happy to have your positive comments. Here are my thoughts.

1. The basic architecture of `Parallelformers` is similar to that of `DeepSpeed`. Now that the integration of HuggingFace `Transformers` and DeepSpeed is in progress, I think it would be the best if we (HuggingFace, MS, and Tunib) all could cooperate together. A day ago I received such a proposal (working with DeepSpeed team) from Reza, an MS engineer.

2. Probably I should be working on the implementation of model parallelization through tensor slicing. Currently Parallelformers supports almost all models in HuggingFace without fused CUDA kernels. If we work together on this project, which I really hope to do, I plan to find a way to integrate the current mechanism to the fused CUDA kernel. I believe if this works out, we can obtain both the speed of the fused CUDA kernel and the scalabilities of Parallelformers.

3. I also think training parallelization is an important issue. In my opinion, it is necessary to consider integrating Tensor MP with DP and DDP. I hope ultimately all the models in Transformers support 3D parallelization through Zero + Pipeline with Tensor MP.

4. Currently it is not possible to deploy a model on the web server with DeepSpeed, which I think a critical issue. Obviously, Parallelformers started to tackle it, but I'm open to any cooperation to find a better solution.

**Comment by stas00** [comment]
Everything you shared sounds good to me, @hyunwoongko.

With regards to 3D parallelism. currently the main obstacle in HF Transformers to support Pipeline Parallelism (PP) is the presence of multiple optional features that prevent the model from being convertable to `nn.Sequential` which is the prerequisite for implementing PP. Though Sagemaker docs claim that they are able to use PP without a model being converted to `nn.Sequential`. So it's possible that to get to PP we may have to make alternative versions stripped of the optional features. But we can discuss this when we are done with TP (tensor parallelism). 

I posted this earlier, could you please address this?

> Let's discuss the training side of it. I think to most users of HF transformers that would be the most important application of Tensor parallelism. So in the deleted note you mentioned that DDP-support needs to be integrated to make it work in training. That's the MPU part, right? And we probably should think about Pipeline too while building the MPU, while not implementing it just yet.

Practically, since you understand your code the best, please let's discuss how to approach the integration of it.

Also let me add a reference to your project at https://huggingface.co/transformers/master/parallelism.html#tensor-parallelism

**Comment by hyunwoongko** [comment]
> With regards to 3D parallelism. currently the main obstacle in HF Transformers to support Pipeline Parallelism (PP) is the presence of multiple optional features that prevent the model from being convertable to nn.Sequential which is the prerequisite for implementing PP. Though Sagemaker docs claim that they are able to use PP without a model being converted to nn.Sequential. So it's possible that to get to PP we may have to make alternative versions stripped of the optional features. But we can discuss this when we are done with TP (tensor parallelism).

I totally agree with your opinion. An interesting thing is that my former colleague was the first to implement PP on a torch. ([torchgpipe](https://github.com/kakaobrain/torchgpipe)) He first implemented it in a way that uses `nn.Sequential`. So, if possible, I'll try to ask him for advice. 

One thing I'm considering is to utilize `nn.ModuleList` in PP. Currently, most of the Transformers models are implemented as `nn.ModuleList`. I think it would be good to use it for PP. the fact that Sagemaker can parallelize Huggingface's models easily means there's something we haven't been able to figure out. I hope that in the future we will continue to work together to find such a scalable way.

> Let's discuss the training side of it. I think to most users of HF transformers that would be the most important application of Tensor parallelism. So in the deleted note you mentioned that DDP-support needs to be integrated to make it work in training. That's the MPU part, right? And we probably should think about Pipeline too while building the MPU, while not implementing it just yet.

Yes, we need to implement training side of it. However, it seems a little difficult to use NVIDIA's mpu implementation in transformers. My idea is to leverage the mechanism of parallelformers again. It is to utilize the most of existing transformers code. When I was implementing parallelformers, I was able to successfully parallelize most models `forward` by changing only a few `nn.Linear` layers while utilize existing transformers codes. And I think this can be applied to backward as well. However, combining this with the fused CUDA kernel on DeepSpeed side can be quite difficult. I think forward is ok, but backward is hard. Because backward is not implemented in their Tensor MP kernel.

Combining DP and DDP probably requires minor changes to the existing torch implementation. As you know, with DP and DDP, same model parameters are broadcast to all GPU. And, each piece of data is sent to each GPUs.

e.g. 
- if bsz=16, n_gpus=2
- gpu1=batch 0-7
- gpu2=batch 8-15 

This needs to be partitioned. If Tensor MP size is 2, we should create two partitions. 

e.g. 
- mp_group1=gpu 0, 1
- mp_group2=gpu 2, 3

And I think that the data should be split by each partition, not by each GPU. 

e.g. 
- if bsz=16, n_gpus=4, mp_size=2
- mp_group1(gpu0,1)=batch 0-7
- mp_group2(gpu2,3)=batch 8-15

**Comment by hyunwoongko** [comment]
I wrote it with a little help from a translator. If you can't understand, please tell me :)

---

Here is a fisrt draft of the collaboration plan. Please feel free to comment. Everyone involved in the collaboration will be able to modify this plan depending on the circumstances.

### Step 1. Implement Parallelformers Tensor MP to DeepSpeed (TUNiB)
The method of replacing the existing layer uses the scalable method of parallelformers. This does not change the entire transformer layer, but a method to replace a few linear layers with a sliced ​​linear layer or a sliced ​​all-reduce linear layer. Since DeepSpeed's Tensor MP replaced the entire Transformer layer, it could not reflect the specific mechanism of each model.

Firstly, I will implement this method and PR to DeepSpeed. (And this is what the DeepSpeed ​​team wants me to do. refer tp [here](https://github.com/microsoft/DeepSpeed/issues/1161#issuecomment-883739436) Ultimately, it's a good idea to archive parallelformers after most of the mechanisms of parallelformers are moved in DeepSpeed. It's a pity that our toolkit will be archived, but I think user accessibility is much more important. Because Parallelformers are less accessible compared to HF Transformers and MS DeepSpeed.

### Step 2. Collaborate TUNiB and DeepSpeed about fused CUDA kernel
However, it is quite challenging to combine it with the CUDA kernel in the training process. In my opinion, it would not be difficult to implement forward pass, but the problem is backward. There is currently no backward pass implementation in the Tensor MP kernel in DeepSpeed. Because Tensor MP is now provided as inferences, they did not need to be implemented backward pass. Unfortunately, since I do not understand the CUDA code at a high level, it will be difficult for me to write the CUDA backward code myself.

Therefore, collaboration with DeepSpeed ​​should be made in this part. It would be nice if we could collaborate with DeepSpeed ​​and discuss about backward implementation of the DeepSpeed ​​Tensor MP kernel. If this is impossible, it may be difficult to use the CUDA kernel during the training process.

### Step 3. Collaborate Huggingface andTUNiB about new fused Tensor MP kernel
In this step, we will add the newly implemented Tensor MP kernel by DeepSpeed and TUNiB into the HuggingFace. I think it will be similar to the [Policy](https://github.com/tunib-ai/parallelformers/tree/main/parallelformers/policies) I implemented in parallelformers. 

There are two methods to add to HuggingFace side.

1) Like `modeling_MODEL.py` and `tokenization_MODEL.py` in each model directory, we can create `parallel_MODEL.py` about the parallelization policy.

2) Alternatively, it is also worth considering about utilizing config.json. However, this can also be a fairly large work because  every config.json files uploaded to hub needs to be changed.

### Step 4. Collaborate Huggingface andTUNiB about DP, DDP, PP
Once Tensor MP is done, we will be able to proceed with combining it with DP and DDP. At the same time, It would be good to consider about implementing PP using nn.ModuleList. In my opinion, the existing PP based on `nn.Sequntial` is not suitable for HF transformers. I will to ask a former colleague for their opinion on a PP implementation based on `nn.ModuleList`. In this process, if we use the pipelining method based on PipeDream Flush implemented in DeepSpeed, collaboration with DeepSpeed will be needed.

**Comment by stas00** [comment]
We probably should discuss PP elsewhere and focus in this thread on what's already working in your project. So I will give a brief overview only:

> I totally agree with your opinion. An interesting thing is that my former colleague was the first to implement PP on a torch. ([torchgpipe](https://github.com/kakaobrain/torchgpipe)) He first implemented it in a way that uses `nn.Sequential`. So, if possible, I'll try to ask him for advice.

Great!

> One thing I'm considering is to utilize `nn.ModuleList` in PP. Currently, most of the Transformers models are implemented as `nn.ModuleList`. I think it would be good to use it for PP. the fact that Sagemaker can parallelize Huggingface's models easily means there's something we haven't been able to figure out. I hope that in the future we will continue to work together to find such a scalable way.

The 3 frameworks that currently provide PP as an API that I know off (fairscale, deepspeed and pytorch's recent versions) all require `nn.Sequential`. So unless we implements a custom PP, `nn.ModuleList` won't do.

Actually, the main complication of the current models, is the inputs/outputs. PP requires simple tensor variables that can be sliced at the batch dimension. HF models have a gazillion of variables that aren't tensors and thus can't be sliced. Some variables are tuples of tuples and are used as aggregates. 

If you'd like to see the sort of jumps through the hoops I had to go through to make it work for t5, please see:
- https://github.com/huggingface/transformers/pull/9765
- https://github.com/huggingface/transformers/pull/9940 (and this attempt I didn't even bother finishing)

Note that over the spring pytorch has developed a much more user-friendlier PP API, which now allows passing non-tensor variables, which should make things much easier.

Most likely we will have to make stripped versions of the current models which support only the features that PP can accommodate.

**Comment by stas00** [comment]
> Yes, we need to implement training side of it. However, it seems a little difficult to use NVIDIA's mpu implementation in transformers. 

I wasn't referring to a specific MPU implementation. Deepspeed has one too. It's basically the manager of all dimensions of parallelism. The only reason I mentioned it so that we consider the future PP dimension as we develop the manager.

> My idea is to leverage the mechanism of parallelformers again. It is to utilize the most of existing transformers code. When I was implementing parallelformers, I was able to successfully parallelize most models `forward` by changing only a few `nn.Linear` layers while utilize existing transformers codes. And I think this can be applied to `backward` as well. However, combining this with the fused CUDA kernel on DeepSpeed side can be quite difficult. I think `forward` is ok, but `backward` is hard. Because backward is not implemented in their Tensor MP kernel.

Then we start with just that.

> Combining DP and DDP probably requires minor changes to the existing torch implementation. As you know, with DP and DDP, same model parameters are broadcast to all GPU. And, each piece of data is sent to each GPUs.
> 
> e.g.
> 
>     * if bsz=16, n_gpus=2
> 
>     * gpu1=batch 0-7
> 
>     * gpu2=batch 8-15
> 
> 
> This needs to be partitioned. If Tensor MP size is 2, we should create two partitions.
> 
> e.g.
> 
>     * mp_group1=gpu 0, 1
> 
>     * mp_group2=gpu 2, 3
> 
> 
> And I think that the data should be split by each partition, not by each GPU.
> 
> e.g.
> 
>     * if bsz=16, n_gpus=4, mp_size=2
> 
>     * mp_group1(gpu0,1)=batch 0-7
> 
>     * mp_group2(gpu2,3)=batch 8-15

Yes, that's the whole point of MPU. DP doesn't even need to know about TP, it just sees gpu0 and gpu2 - it has no idea there are more GPUs in the pipe. Each parallel dimension typically hides its existence from other dimensions, which allows things to keep simple.

**Comment by stas00** [comment]
Your collaboration plans is very clear, @hyunwoongko.

Thank you for your inspiration to share your work for the good of all! It's true that being part of a "bigger pie" will make your work accessible to a lot more users.

Wrt step2, you know that Deepspeed has a fully TP implementation, except not in CUDA kernels - perhaps this can be utilized instead?

Otherwise please ping or tag me when you need my input here or on the Deepspeed github.

Looking forward to this inspiring collaboration, @hyunwoongko

**Comment by hyunwoongko** [comment]
First of all, we need to discuss this collaborative process with @@RezaYazdaniAminabadi. 
Can we discuss it here? I'm curious about your opinion.

**Comment by hyunwoongko** [comment]
> Wrt step2, you know that Deepspeed has a full TP implementation, except not in CUDA kernels - perhaps this can be utilized instead for backward?

I'll review the code soon. Thank you.

**Comment by hyunwoongko** [comment]
https://tunib.notion.site/TECH-2021-07-26-Parallelformers-Journey-to-deploying-big-models-32b19a599c38497abaad2a98727f6dc8

Here is the English version of the blog post!

**Comment by hyunwoongko** [comment]
@stas00 Sorry for the delay this work. We are also making a public large-scale model that can cover Asian languages. I've been very busy these days, so I haven't had much time to contribute to Hugging Face. I will work on it as possible as soon.

**Comment by stas00** [comment]
Also pinging @siddk whose team also has been working on improving `transformers` to support TP/PP https://github.com/stanford-crfm/mistral.

For context, while your team was on a summer break, @hyunwoongko implemented Parallelformers and we started discussing how to integrate their work, while planning integration of Deepspeed CUDA kernels for TP.

So now that your team is getting back let's discuss how to best collaborate.

**Comment by siddk** [comment]
Oh this is awesome, thanks @stas00 and nice to meet you @hyunwoongko. Let me get up to speed on this thread, but this looks like amazing work!

**Comment by hyunwoongko** [comment]
@siddk Hello. Could you please explain so I can get the context? :)

**Comment by hyunwoongko** [comment]
I will resume this work from this weekend. Since my company is so busy now, most of the open source work will probably be done on weekends. I will working on deepspeed this week. I had an offline meeting with them and we are discussing how to combine. (Probably integration with Huggingface transformers will not take place soon because it is steps 3 and 4.)

**Comment by jaketae** [comment]
It's really cool to see this collaboration in the pipeline! I'm not affiliated with any of the frameworks/organizations here at stake, but I do come from HF BigScience side of things where I've briefly discussed things with @stas00. If there's grunt work or anything else that has to be done, I'd be more than happy to contribute in ways that I can.

**Comment by hyunwoongko** [comment]
@jaketae I already know you by KoClip. nice to meet you. Your help would be of great help. :)

@stas00 Currenlty, we need to talk more with the DeepSpeed team. I will first integrate the parallelformers features into deepspeed. However, what deepspeed and transformers currently want is slightly different, so you may need to adjust it.

1) deepspeed wants to improve deepspeed-inference, maybe they are not considering training features.
2) transformers want to training features with 3D parallelization. (and as we said before, we have to consider a megatron-lm style mpu if we implement training features with pp and dp).

**Comment by stas00** [comment]
As I commented in another issue:

HF transformers wants both training and inference. It's just that we have a lot more users using the library for training. So there is definitely not misalignment between the two.

Remember that Deepspeed already has TP and PP, so they are just missing inference.

HF Transformers doesn't have those yet, hence the difference.

**Comment by stas00** [comment]
> @siddk Hello. Could you please explain so I can get the context? :)

https://twitter.com/siddkaramcheti/status/1430195543301492744

**Comment by stas00** [comment]
> If there's grunt work or anything else that has to be done, I'd be more than happy to contribute in ways that I can.

@jaketae, the idea is to first pick one model and port it to TP and later PP. Then we will have to replicate this for all models (or at least models that will support this), so there will be a ton of work for quite a few people to contribute.

**Comment by hyunwoongko** [comment]
I will close this issue. lets discuss in https://github.com/huggingface/transformers/issues/13690
